[PL] [134362869] Handle cascading versions

pasha opened the pull request:
[Support cascading lookup](https://www.pivotaltracker.com/story/show/134362869)

- add rubocop and fix errors
- added option to allow cascading per version

Commit:
[PL] [134362869] Handle cascading versions

https://www.pivotaltracker.com/story/show/134362869

- add rubocop and fix errors
- updated Readme
- updated gem version

mirineumark commented on lib/phenotype/call_handler.rb:
```diff
@@ -0,0 +1,29 @@
+    response[0].is_a?(Numeric) && response[1].is_a?(Hash) && response[2].is_a?(Array)
+  end
+
+  def status?
```
I think this name does not describe what it does very well. Maybe `not_found?`

pasha replied:
updated method name to `not_found?` 👍

mirineumark commented on lib/phenotype/version.rb:
```diff
@@ -1,3 +1,3 @@
 module Phenotype
-  VERSION = '0.2.0'
+  VERSION = '0.3.0.beta1'.freeze
```
Why beta? Why not just bump to 0.3.0? I think the below 1.0.0 implies beta.

pasha replied:
so that final will get `0.3.0` just in case if we see any issues and need to create few more beta versions.

mirineumark commented on lib/phenotype/versioned_app.rb:
```diff
@@ -1,48 +1,25 @@
-
-    def default_response(status_code)
-      Proc.new { [status_code, { 'Content-Type' => 'text/html', 'Content-Length' => "#{status_code}".size.to_s}, ["#{status_code}"]] }
+      Versioner.new(routes: routes, strategies: strategies, env: env).call
```
Nice refactor 👍

mirineumark commented on spec/call_handler_spec.rb:
```diff
@@ -0,0 +1,51 @@
+      let(:response) { [404, { 'X-Cascade' => 'pass' }, ['sucess']] }
+      subject { handler.retry? }
+
+      it 'be true' do
```
You should make this a one liner.

``` ruby
it { is_expected.to be_truthy }
```

mirineumark commented on spec/call_handler_spec.rb:
```diff
@@ -0,0 +1,51 @@
+      subject { handler.retry? }
+
+      it 'be false' do
+        expect(subject).to be_falsey
```
Also a one liner:

``` ruby
it { is_expected.to be_falsey }
```

mirineumark commented on spec/call_handler_spec.rb:
```diff
@@ -0,0 +1,51 @@
+      let(:response) { [200, { 'X-Cascade' => 'pass' }, ['sucess']] }
+      subject { handler.retry? }
+
+      it 'be false' do
```
Also a one liner:

``` ruby
it { is_expected.to be_falsey }
```

mirineumark commented on spec/versioner_spec.rb:
```diff
@@ -0,0 +1,174 @@
+          end
+
+          it 'be true' do
+            expect(subject).to eql(error_response)
```
Also a one liner:

``` ruby
it { is_expected.to eql error_response }
```

mirineumark commented on spec/versioner_spec.rb:
```diff
@@ -0,0 +1,174 @@
+          end
+
+          it 'be true' do
+            expect(subject).to eql(ok_response)
```
Also a one liner:

``` ruby
it { is_expected.to eql ok_response }
```

mirineumark commented on spec/versioner_spec.rb:
```diff
@@ -0,0 +1,174 @@
+
+      context 'and version matches existing route call' do
+        it 'returns 200 response' do
+          expect(subject).to eql(ok_response)
```
Also a one liner:

``` ruby
it { is_expected.to eql ok_response }
```

mirineumark commented on spec/versioner_spec.rb:
```diff
@@ -0,0 +1,174 @@
+
+    context 'When Errors exist' do
+      it 'returns 406 error' do
+        expect(subject.first).to eql(406)
```
Also a one liner:

``` ruby
its(first) { is_expected.to eql 406 }
```

pasha replied:
do not want to introduce another gem just for this matcher, updated `subject` to include `.first`

mirineumark reviewed (changes requested):
A few small things, but looks good overall.

rkledesma commented on lib/phenotype/call_handler.rb:
```diff
@@ -0,0 +1,29 @@
+  end
+
+  def rack_response?
+    response[0].is_a?(Numeric) && response[1].is_a?(Hash) && response[2].is_a?(Array)
```
You could use `first` `second` `third` Array methods here

pasha replied:
these will be available in rails but not ruby

```
~/source/phenotype on pl/add/cascading_fetch ✗ $ pry
bundle exec pry
ruby-2.2.3-p173 (main):0 > a = [1, 2, 3]
=> [1, 2, 3]
ruby-2.2.3-p173 (main):0 > a.first
=> 1
ruby-2.2.3-p173 (main):0 > a.second
NoMethodError: undefined method `second' for [1, 2, 3]:Array
from (pry):3:in `__pry__'
ruby-2.2.3-p173 (main):0 > a.third
NoMethodError: undefined method `third' for [1, 2, 3]:Array
from (pry):4:in `__pry__'
```

colinrymer commented:
So I'm pretty opposed to the idea that this PR introduces. I think falling back to a previous version of the API is just asking for problems, e.g.:
- How do I ensure my changes to a prior version don't affect the newer versions in a problematic manner?
- How do I handle endpoints that I no longer want to respond to?
- I'm introducing additional work/cognitive overhead - now if I want to see the code for a given version's endpoint, I have to look in the given version, then through each prior version until I find the code I'm looking for.

Instead, if we introduce a new version of an API that has near identical endpoints, we should just copy the old version into the new one. While we're duplicating code, I'm totally okay with it in this scenario, because we're preferring explicit over implicit. We should basically be treating each version as a new codebase, IMO.

pasha commented:
I am partially agreeing with you, but I think any code duplications could introduce more issues.
- This will increase how long specs run especially on larger projects
- If you need to fix some issue in later version which is also an issue in previous versions, you will need to fix it in multiple places.
- To handle no longer supported end point simply return 404 from them(this will not cause additional lookup in previous versions) - not ideal but much less duplication.
- I think it's easier to look through multiple versions when code doesn't get duplicated, if `v3/some/route` doesn't exist you simply look in previous version.

I don't think there is a silver bullet for how to version apps, i just don't like duplicated code, i think it's prone to errors.

rkledesma commented on spec/call_handler_spec.rb:
```diff
@@ -0,0 +1,45 @@
+      it { is_expected.to be_falsey }
+    end
+
+    context '200 and X-Cascad' do
```
Some typos here: should be 'Cascade' and 'success'

pasha replied:
fixed

rkledesma commented on spec/versioned_app_spec.rb:
```diff
@@ -43,31 +43,38 @@ def version(env)
+    context 'with 0 strategies' do
+      subject { described_class.new.call(env) }
+      it 'returns a 406' do
+        expect(subject[0]).to eql(406)
```
might as well use `first` in this file

pasha replied:
refactored

rkledesma commented on spec/versioner_spec.rb:
```diff
@@ -0,0 +1,167 @@
+      end
+    end
+
+    context 'To Many Strategies' do
```
'To' should be 'Too'

pasha replied:
fixed

rkledesma commented:
I'm inclined to agree with @colinrymer -- the same thoughts came into my mind as well, particularly about explicitness and cognitive load.

Commit:
[PL] [134362869] Added support controlling cascade

- support for each version to allow cascade down
- rename classes/files
- added name-spacing

Signed-off-by: Pasha Lifshiz <plifshiz@rentpath.com>

pasha commented:
build ci

pasha commented:
build ci

pasha commented:
build ci

pasha commented:
build ci

pasha commented:
build ci

pasha commented:
build ci

pasha commented:
build ci

pasha commented:
build ci

dillo commented:
👍
